contracts/contracts/OpenAiChatGpt.sol

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+// Uncomment this line to use console.log
+// import "hardhat/console.sol";
+
+interface IOracle {
+    struct OpenAiRequest {
+        string model;
+        int8 frequencyPenalty;
+        string logitBias;
+        uint32 maxTokens;
+        int8 presencePenalty;
+        string responseFormat;
+        uint seed;
+        string stop;
+        uint temperature;
+        string tools;
+        string toolChoice;
+        string user;
+    }
+
+    struct OpenAiResponse {
+        string id;
+
+        string content;
+        string functionName;
+        string functionArguments;
+
+        uint64 created;
+        string model;
+        string systemFingerprint;
+        string object;
+
+        uint32 completionTokens;
+        uint32 promptTokens;
+        uint32 totalTokens;
+    }
+
+    function createOpenAiLlmCall(
+        uint promptId,
+        OpenAiRequest memory request
+    ) external returns (uint);
+
+    function createFunctionCall(
+        uint functionCallbackId,
+        string memory functionType,
+        string memory functionInput
+    ) external returns (uint i);
+}
+
+contract OpenAiChatGpt {
+
+    struct Message {
+        string role;
+        string content;
+    }
+
+    struct ChatRun {
+        address owner;
+        Message[] messages;
+        uint messagesCount;
+    }
+
+    mapping(uint => ChatRun) public chatRuns;
+    uint private chatRunsCount;
+
+    event ChatCreated(address indexed owner, uint indexed chatId);
+
+    address private owner;
+    address public oracleAddress;
+
+    event OracleAddressUpdated(address indexed newOracleAddress);
+
+    IOracle.OpenAiRequest private config;
+
+    constructor(address initialOracleAddress) {
+        owner = msg.sender;
+        oracleAddress = initialOracleAddress;
+        chatRunsCount = 0;
+
+        config = IOracle.OpenAiRequest({
+        model : "gpt-4-turbo-preview",
+        frequencyPenalty : 21, // > 20 for null
+        logitBias : "", // empty str for null
+        maxTokens : 1000, // 0 for null
+        presencePenalty : 21, // > 20 for null
+        responseFormat : "{\"type\":\"text\"}",
+        seed : 0, // null
+        stop : "", // null
+        temperature : 10, // Example temperature (scaled up, 10 means 1.0)
+        tools : "[{\"type\":\"function\",\"function\":{\"name\":\"web_search\",\"description\":\"Search the internet\",\"parameters\":{\"type\":\"object\",\"properties\":{\"query\":{\"type\":\"string\",\"description\":\"Search query\"}},\"required\":[\"query\"]}}}]",
```

Comment:
OpenAI format tool description
https://platform.openai.com/docs/api-reference/chat/create
Names and params have to match the tools our oracle supports

Reply:
I wonder - if we only support using our tools anyway, then the developer actually faces a binary choice per-tool: should I enable this tool or not.

So maybe a better and much simpler interface would be that the developer adds a list of strings as tools -- and we populate the actual interface. I realize this is a departure from the OpenAI API, but it would seem to reduce work for the developer a lot?

Reply:
That would make sense and I considered it, but this leaves the prompting on the oracle side, imo it makes sense to give devs the possibility to prompt theirs tools how they see fit, depending on the use case? But yes this format is not optimal :/

Reply:
It doesn't leave prompting on oracle though? The only thing the oracle does is rewrite this list:
```
[web_search, blabla_tool]
```

into
```
[{web_search_long_ass_json}, {blabla_tool_long_ass_json}, 
```

Reply:
This is the format that is used in OpenAI (and this contract example)
```
[
  {
    "type": "function",
    "function": {
      "name": "web_search",
      "description": "Search the internet",
      "parameters": {
        "type": "object",
        "properties": {
          "query": {
            "type": "string",
            "description": "Search query"
          }
        },
        "required": [
          "query"
        ]
      }
    }
  }
]
```
If we just had tool names here then oracle would have to add the descriptions, which might change depending on the use case